### ISSUE TITLE
User can add server credentials when adding project

### ISSUE THREAD

**Issue by seadowg** [getodk/collect#4428]
~~Blocked by #4422~~
Mockup [here](https://www.figma.com/file/nDsMAYQDIflyl7rznEhwnN/ODK-Collect-Multiple-Project-Support-(v2021.2)?node-id=37%3A219)

### Acceptance

- [x] Given I'm adding a project by manually entering details
When enter a URL
And a username/password
And click "Add"
Then the project shows the username and server host under the project name
And the project has a generated name (see notes)
And the project has a generated icon (see notes)

- [x] Given I'm adding a project by manually entering details
When enter a URL
And I don't enter a username or password
And click "Add"
Then the project shows the server host under the project name
And the project has a generated name (see notes)
And the project has a generated icon (see notes)

### Notes

* Server host should show without protocol, path or query but with subdomain so `https://www.example.com/something?id=123` would show as `www.example.com`.
* The auto-generated project name should be everything in the host name up until the path. For example, `https://www.example.com/something?id=123` would be `www.exampe.com`.
* The auto-generated project icon would just be the first letter of the project name.
* For the moment the project color can be anything. Probably best to just use the default blue we've been using for demo/existing.

**Comment by seadowg** [comment]
@grzesiek2010 right now on master it doesn't look like the settings change when switching between projects - if you change the server in one project and then switch to another you'll see the same server for example. That's fine as that's what I'm working on in #4544, so I've just changed the acceptance here slightly to focus more on the UI around entering project details and then seeing those details in the project list (which will work using `SettingsProvider#getGeneralSettings` with the project ID I imagine).

**Comment by grzesiek2010** [comment]
I think it must be something small because when you switch projects we save `CURRENT_PROJECT_ID` in meta settings then whenever `SettingsProvider` is used it takes that current project id to return proper settings. That should be enough...

**Comment by grzesiek2010** [comment]
ahhh taht's because `providesGeneralSettingsStore` and `providesAdminSettingsStore` in `AppDependencyModule` are singletons, that's the bug.

**Comment by seadowg** [comment]
@grzesiek2010 ah yeah! Looks like I made that change in #4544: https://github.com/getodk/collect/pull/4544/commits/33126eb1ae7c7d1cc41757746e27ed69decf1264#diff-327c65e40a12e80150d6fd6ca4113fc600545c80e650202fdb633e84bd9c32d7L493. That PR can take care of it then.

**Comment by lognaturel** [comment]
I'm sorry, I didn't read this carefully previously. Requirements from [the forum discussion](https://forum.getodk.org/t/collect-multiple-projects-support/23340/2) mention defaulting project name and character if they're not set but they don't say anything about visible defaults when filling out the manual configuration form. I don't think they should default in the form. In most cases, that would be annoying because the user would want to set a nice project name and pick their own icon. Apologies, that should have been explicit. Does that make sense to you @grzesiek2010, @seadowg or do you think there's value in autopopulating the visible fields? Did we discuss doing it that way previously?

My expectation was that URL would be the only required field when configuring a server. Username and password would be optional and left blank if not populated. Project name, Project icon and Project color would be optional and auto-populated if left blank. Again, I realize that wasn't explicit.

When a user saves the form, any of project name, icon and color that are missing are populated according to the rules we'd described (with examples at https://github.com/getodk/collect/issues/4579#issuecomment-848114962). The exact same thing happens when a QR code is scanned or on upgrade (with the additional requirement that `demo.getodk.org` has a special name of "Demo" but I don't mind dropping that):
- Project name it is populated with the full domain name including subdomains but no path 
- Project icon is populated with the first character of the project name (note project name could have been user-specified or auto generated based on the rule above, it doesn't matter)
- Project color is computed from the profile name 

Hopefully this behavior can be shared between the three cases where it's needed.

Is the first bullet about showing `username / domain` in the projects list dialog?

**Comment by lognaturel** [comment]
> The auto-generated project name should be everything in the host name up until the first .

Here are some examples of why I'd written it up as the full domain name:
- `https://www.example.com` => `www`
- `https://odk.mycompany.com` => `odk`
- `https://central.mycompany.com` => `central`
- `https://mycompany.getodk.cloud` => `mycompany` (this might look reasonable but imagine a company with 6 very different projects. Showing Mycompany as the project name doesn't give the clue that maybe project name is configurable)
- `https://test.getodk.cloud`, `https://test.getodk.org`, `https://test.mydomain.com` => `test`
- `https://kf.kobotoolbox.org/` => `kf`

**Comment by seadowg** [comment]
> I don't think they should default in the form. In most cases, that would be annoying because the user would want to set a nice project name and pick their own icon. Apologies, that should have been explicit. Does that make sense to you @grzesiek2010, @seadowg or do you think there's value in autopopulating the visible fields? Did we discuss doing it that way previously?

@lognaturel From the conversations @grzesiek2010 and I had we were thinking that the name appears in the field, but you can still edit it, and if you do edit it, it will not again be autofilled. Hopefully that's clear in the acceptance. So they can still set their own name and icon if they choose. I don't feel strongly that that is required as I could see that we think about this as the secondary path for configuration, but it does feel like it adds convenience for those who are copy-pasting a URL and just want to get setup as quickly as possible. I'm happy to just drop autofill in the form if both you and @grzesiek2010 agree it's easier not to do it!

> Here are some examples of why I'd written it up as the full domain name

Happy to change it to just the full domain name. I'm not really convinced one or the other is better as it feels likely to me that the subdomain will be where the "unique" part is and if not, it'll be in the path. Given it's editable I don't think it's a big deal to just do the simplest thing here. I'll change the acceptance. @grzesiek2010 let me know if that's going to be really awkward to unwind and if so I can make another issue for it.

> Project color is computed from the profile name

Lets leave color for #4493 right now. I think it's easier to just use a "default" everywhere (like we are now) and then come back for generated/picked color in one go.

> Is the first bullet about showing username / domain in the projects list dialog?

Yup!

**Comment by seadowg** [comment]
From @lognaturel:

> I've been meditating on all this after writing up https://github.com/getodk/collect/issues/4428#issuecomment-848261115 and thinking about what I said at https://github.com/getodk/collect/issues/4579#issuecomment-848114962 about most users not changing the default way a project is presented. 

> Running this PR made me realize I made a mistake by not mocking the screen up with the keyboard. Without the keyboard, server info + project display configuration is somewhat digestible, and I didn't realize that it's a lot of prompts without context. I believe that @grzesiek2010 you chose not to pop up the keyboard when the screen first loads to give users a sense of what can be configured. That makes sense but it does mean an extra tap into the URL field which feels unexpected and awkward to me. And then depending on the device, the keyboard likely hides fields that get auto-populated which is extra confusing once the user scrolls down to see them. Or if you're unlucky like me, the keyboard exposes exactly the first three fields and it's very easy to forget there's more below.

> I also think you may have decided to do the auto-population of the fields in an attempt to provide some context on what the project configuration values should be. That helps but it still feels surprisingly confusing to me. Where will the name be shown? The icon? Why a color? I said most users won't change these defaults and I really believe they won't have any reason to.

> This was very informative to actually try. Having seen it, I propose we only ask for URL/username/password. That's similar to what e.g. Discourse does -- when you create a new account, the first letter of your username is used as your avatar along with some color and then you can choose to configure your avatar differently in settings if you want. That seems like a better user experience. I also mocked it up [here](https://www.figma.com/file/nDsMAYQDIflyl7rznEhwnN/ODK-Collect-Multiple-Project-Support-(v2021.2)?node-id=37%3A219) and another advantage is that it gives us a lot more options for Google Drive config, I think.

> Happy to hop on a call if y'all aren't sure about this change and/or what I described around fallback/default values.
